### ISSUE TITLE
Map#cast has unexpected null behavior

### ISSUE THREAD

**Issue by trinarytree** [dart-lang/sdk#53818]
this
```dart
  final l = <int?>[1, null];
  final l2 = l.cast<int>();
  l2[1];
```
throws.  yay.  but this 
```dart
  final m = <int, int?>{1: null};
  final Map<int, int> m2 = m.cast<int, int>();
  m2[1];
```
doesn't, which i find confusing.  my theory for why it doesn't throw:
`operator []` is allowed to return a null to indicate that
the key wasn't found OR that the value is null.
this is a weird case because we have a map with a non-nullable value type,
yet `operator []` is returning null when the key _is_ found.

i feel like this perhaps should be considered a bug in `Map` - the above access probably should throw.

an example of how _not_ throwing could lead to bugs:
```dart
/// Returns the number of [people] who are non-citizens.
///
/// Assumes the keys of [citizenToSsn] are the citizens.
int countNonCitizens(
    Iterable<Person> people, Map<Person, int> citizenToSsn) {
  var numNonCitizens = 0;
  for (final person in people) {
    final ssn = citizenToSsn[person];
    if (ssn == null) {
      // Conclude that person is not a citizen.
      numNonCitizens++;
    }
  }
  return numNonCitizens;
}
```
the above code may not be maximally elegant, but it doesn't look wrong either, yet it is wrong.
with current semantics, the deduction (`ssn == null` implies `person` is not a key) is invalid.

\> dart --version
Dart SDK version: 3.0.7 (stable) (Mon Jul 24 13:17:56 2023 +0000) on "macos_x64"

**Comment by lrhn** [comment]
You're making a good point. It really should be throwing.

It requires more work than just `V2? operator[](K2 key) => _map[key as K1] as V2?;`, but it'd be more consistent and makes the following pattern behave as expected: 
```dart
extension <K, V> on Map<K, V> {
  R mapLookup<R>(K key, R Function(V) onNonNullValue, R Function() onNullValue, R Function() onAbsent) {
    V? value = map[key];
    if (value != null) {
      return onNonNullValue(value);
    }
    if (map.containsKey(key)) {
      return onNullValue();
    }
    return onAbsent();
  }
}
```
This code looks correct, for any valid implementation of the `Map` interface, but the `cast`-map which casts values to non-`null`, will fall into the second branch without throwing. 
Passing in a `Map<int, int>` and getting the `onNullValue` callback is definitely confusing.

It's *consistent* that `containsKey` says `true`, but lookup throws. Not great for the `Map` API, it really shouldn't happen, but consistent.
But it's not really OK that `[]` doesn't throw. I worry a little about the implementation of patterns, which use the "`[]` + `containsKey`-if-`null`" pattern. The spec for that should be written to handle invalid implementations without crashing (just not necessarily with a useful result, that's what you get for doing an invalid implementation), but then we shouldn't have invalid implementations in the platform libraries.

But it is efficient, and it shouldn't matter for the *intended* use-case of the `cast` method.
The intended use-case is that you have a map where you *know* that all the values are, say, `Strings`, even though the actual value type is `Object`. Then you do `.cast<K, String>()`, and you can use the map as a `Map<K, String>`, and it never throws because you know what you're doing.

It's not intended to be used to cast a map containing non-`String`s to have a `String` type. That's why it uses `as` casts, which are also intended for "I know what I'm doing, even if the type system can't see it" casting, and doesn't provide, say, an error handler that can convert invalid values to valid values.

All in all, if there is no problem with pattern implementations, I'll prefer to not change the implementation.

We just have to document, even more explicitly, that if `cast` is used in a way other than it's intended use, where all values actually have the cast-to type, the behavior may not satisfy the `Map` interface contract. 
Throwing in `[]` doesn't satisfy the contract either, so it's not really new, its just a new and different way of failing to do so.

(Honestly, I'd prefer to just remove the `cast` methods again. They were introduced with the stricter type system of Dart 2.0, mostly to allow existing code to more easily be migrated. If you had a `List<Object>` containing just `String`s in Dart 1, you could just use it as a `List<String>` at runtime, at least in production mode. Types were only checked in debug mode.
And we've come some way since Dart 2.0, with compilers actually using the sounder type system to omit unnecessary type checks. Because of that, it's not always more efficient to use `cast` than to just create a new map with the new type, even if it copies all the entries, because the `cast` map does type casts/checks on every key and value which occurs as argument or return value, and dynamic type checks with a type parameter are hard to optimize.
And we have *two* entries in the style guide saying to not use it: https://dart.dev/effective-dart/usage#dont-use-cast-when-a-nearby-operation-will-do and https://dart.dev/effective-dart/usage#avoid-using-cast .)

**Comment by trinarytree** [comment]
> The intended use-case is that you have a map where you know that all
> the values are, say, Strings, even though the actual value type is Object

is this really true?  consider this case:

```dart
// We know full well that personToSsn contains men, but
// we promise not to access them.
final womanToSsn = personToSsn.cast<Woman, int>();
for (final woman in women) {
  final ssn = womanToSsn[woman];
  offerAmazingDealOnWomensClothing(woman, ssn);
}
```

i'm not saying i would do it this way, but the docs make it sound like
this is a legit way of using `cast`, so i'm hesitant to say
it assumes all (key, value) pairs conform to the cast types.
perhaps one could say it assumes all _read_ (key, value) pairs conform to the cast types.

btw, i notice that in the first example,
```dart
m2.putIfAbsent(1, () => 2);
```
throws, so at least this method doesn't confuse key absence and null value.

**Comment by lrhn** [comment]
> is this really true?

Yes. That was the intended use-case that the `cast` method was added for.

Limiting which keys are accepted was not considered. Also, it doesn't work since lookup has argument type `Object?`, and doesn't cast the key.

The cats performed are those necessary to ensure the new typing, assuming that the existing keys and values have that type. If not, you may see inconsistent behavior. If you want to avoid that, you should create a new map with only entries that would satisfy need types, like
```dart
{for (var entry in map.entries)
    If (entry case MapEntry(: K2 key, : V2 value)) 
        key: value
}
```
